src/FloatingMenuPlugin.ts

```diff
@@ -0,0 +1,570 @@
+      // If not JSON, treat as plain text
+      tr = view.state.tr.insertText(
+        text,
+        view.state.selection.from,
+        view.state.selection.to
+      );
+    }
+
+    view.dispatch(tr.scrollIntoView());
+  } catch (err) {
+    console.error('Clipboard paste failed:', err);
+  } finally {
+    if (plugin._popUpHandle?.close) {
+      plugin._popUpHandle.close(null);
+      plugin._popUpHandle = null;
+    }
+  }
+}
+
+export async function pasteAsReference(
+  view: EditorView,
+  plugin: FloatingMenuPlugin
+) {
+  try {
+    if (!view.hasFocus()) view.focus();
+    const text = await navigator.clipboard.readText();
+    const parsed = JSON.parse(text);
+    const sliceModel: SliceModel = parsed.sliceModel;
+
+    if (!plugin.sliceManager?.createSliceViaDialog) {
+      throw new Error(
+        'SliceManager or createSliceViaDialog is not initialized'
+      );
+    }
+
+    const val = await plugin.sliceManager.createSliceViaDialog(sliceModel);
+    if (!val) {
+      return;
+    }
+    insertReference(
+      view,
+      val.id,
+      val.source,
+      view['docView']?.node?.attrs?.objectMetaData?.name
+    );
+  } catch (err) {
+    console.error('Failed to paste content or create slice:', err);
+  } finally {
+    if (plugin._popUpHandle?.close) {
+      plugin._popUpHandle.close(null);
+      plugin._popUpHandle = null;
+    }
+  }
+}
+
+export async function pasteAsPlainText(
+  view: EditorView,
+  plugin: FloatingMenuPlugin
+) {
+  try {
+    if (!view.hasFocus()) view.focus();
+
+    const text = await navigator.clipboard.readText();
+    let plainText = text;
+
+    try {
+      const parsed = JSON.parse(text);
+      const slice = Slice.fromJSON(view.state.schema, parsed);
+
+      const frag = slice.content;
+      plainText = '';
+      frag.forEach((node) => {
+        plainText += node.textContent + '\n';
+      });
+      plainText = plainText.trim();
+    } catch {
+      // Not JSON → just keep as is
+    }
+
+    const { state } = view;
+    const tr = state.tr.insertText(
+      plainText,
+      state.selection.from,
+      state.selection.to
+    );
+    view.dispatch(tr.scrollIntoView());
+  } catch (err) {
+    console.error('Plain text paste failed:', err);
+  }
+
+  if (plugin._popUpHandle?.close) {
+    plugin._popUpHandle.close(null);
+    plugin._popUpHandle = null;
+  }
+}
+
+export async function clipboardHasProseMirrorData(): Promise<boolean> {
+  try {
+    const text = await navigator.clipboard.readText();
+    if (!text) return false;
+    const parsed = JSON.parse(text);
+    return !!(
+      parsed &&
+      typeof parsed === 'object' &&
+      parsed.content &&
+      (Array.isArray(parsed.content) || parsed.content.type)
+    );
+  } catch {
+    return false;
+  }
+}
+
+// --- Decoration function ---
+export function getDecorations(doc: Node, state: EditorState): DecorationSet {
+  const decorations: Decoration[] = [];
+
+  doc?.forEach((node: Node, pos: number) => {
```

Comment:
Sonar: Use `for…of` instead of `.forEach(…)`.

You can NOSONAR this rule for Sonar fragments until Sonar fixes this typing bug.
```suggestion
  doc?.forEach( // NOSONAR not an iterable
      (node: Node, pos: number) => {
```